### PR TITLE
refactor: move C{ActiveMasternode, DeterministicMN}Manager, CMasternode{MetaMan, Sync} to NodeContext

### PR DESCRIPTION
**Pull request by kwvg**
## Additional Information

* Depends on https://github.com/dashpay/dash/pull/5966

* `CTxMemPool`'s ProTx logic took the presence of `CDeterministicMNManager` for granted. To account for `CTxMemPool` instances that aren't seeded with the pointer to the manager (usually because the unit test doesn't call for them), refactoring was done to make ProTx code paths conditional on the manager pointer being set to something.
  * Setting the manager pointer is done through `CTxMemPool::Init()`, to avoid having to pass another `std::unique_ptr` const-ref and also because `CTxMemPool` can _technically_ work without the manager if ProTx logic is not needed (and `CTxMemPool::existsProviderTxConflict()` isn't called)

    * `CTxMemPool::Init()` doesn't allow overwriting a not-`nullptr` manager pointer. If that is desired, then `CTxMemPool::Reset()` should be called first. This was done to avoid unintentional double-initialization/overwriting.

## Breaking Changes

None. Changes are limited to refactoring.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation **(note: N/A)**
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

File: src/governance/governance.cpp [hunk 1]
```diff
@@ -393,12 +392,12 @@ void CGovernanceManager::UpdateCachesAndClean()
 
         int64_t nTimeSinceDeletion = nNow - pObj->GetDeletionTime();
 
-        LogPrint(BCLog::GOBJECT, "CGovernanceManager::UpdateCachesAndClean -- Checking object for deletion: %s, deletion time = %d, time since deletion = %d, delete flag = %d, expired flag = %d\n",
-            strHash, pObj->GetDeletionTime(), nTimeSinceDeletion, pObj->IsSetCachedDelete(), pObj->IsSetExpired());
+        LogPrint(BCLog::GOBJECT, "CGovernanceManager::%s -- Checking object for deletion: %s, deletion time = %d, time since deletion = %d, delete flag = %d, expired flag = %d\n",
+            __func__, strHash, pObj->GetDeletionTime(), nTimeSinceDeletion, pObj->IsSetCachedDelete(), pObj->IsSetExpired());
```
Comment (knst) [inline src/governance/governance.cpp:400]: nit: After https://github.com/dashpay/dash/pull/5974/commits/dca7ae20eb57be0656e247dbb2701034488086cc get merged no need more `__func__` here.

So far as function is not actually removed and it still has a correct name, consider removing this refactoring from PR.

File: src/llmq/chainlocks.cpp [hunk 2]
```diff
@@ -25,17 +25,17 @@ namespace llmq
 {
 std::unique_ptr<CChainLocksHandler> chainLocksHandler;
 
-CChainLocksHandler::CChainLocksHandler(CChainState& chainstate, CConnman& _connman, CMasternodeSync& mn_sync, CQuorumManager& _qman,
+CChainLocksHandler::CChainLocksHandler(CChainState& chainstate, CConnman& _connman, CQuorumManager& _qman,
                                        CSigningManager& _sigman, CSigSharesManager& _shareman, CSporkManager& sporkman,
-                                       CTxMemPool& _mempool) :
+                                       CTxMemPool& _mempool, const CMasternodeSync& mn_sync) :
```
Comment (knst) [inline src/llmq/chainlocks.cpp:30]: nit: :cry:

File: src/llmq/utils.h [hunk 3]
```diff
@@ -19,9 +19,11 @@ class CBlockIndex;
 class CDeterministicMN;
 class CDeterministicMNList;
 class CDeterministicMNManager;
-using CDeterministicMNCPtr = std::shared_ptr<const CDeterministicMN>;
+class CMasternodeMetaMan;
 class CSporkManager;
 
+using CDeterministicMNCPtr = std::shared_ptr<const CDeterministicMN>;
```
Comment (knst) [inline src/llmq/utils.h:25]: nit: that's still forward declaration, not sure that empty line here is reasonable but nevermind
```suggestion
using CDeterministicMNCPtr = std::shared_ptr<const CDeterministicMN>;
```

**Review by knst — DISMISSED** [review]
LGTM f967c1648a919a28bfb108c84c18cbcad3bea04e

(CI is not finished yet)

File: src/txmempool.cpp [hunk 4]
```diff
@@ -615,6 +623,23 @@ void CTxMemPool::removeUnchecked(txiter it, MemPoolRemovalReason reason)
     } else
         vTxHashes.clear();
 
+    if (::deterministicMNManager) {
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/txmempool.cpp]: This is done intentionally. The complementing condition check before calling `removeUncheckedProTx` ([source](https://github.com/dashpay/dash/commit/e6088d4e5f8f4c930ca66ad579237df733b35bc2#diff-c065d4cd2398ad0dbcef393c5dfc53f465bf44723348892395fffd2fb3bac522R626-R628)) exists for `addUncheckedProTx` ([source](https://github.com/dashpay/dash/commit/e6088d4e5f8f4c930ca66ad579237df733b35bc2#diff-c065d4cd2398ad0dbcef393c5dfc53f465bf44723348892395fffd2fb3bac522R418-R419)).

The reason why `removeProTxSpentCollateralConflicts` uses an assertion is that `removeProTxSpentCollateralConflicts` is only ever called by `removeProTxConflicts`, which is used conditionally in `CTxMemPool` ([source](https://github.com/dashpay/dash/commit/e6088d4e5f8f4c930ca66ad579237df733b35bc2#diff-c065d4cd2398ad0dbcef393c5dfc53f465bf44723348892395fffd2fb3bac522R986-R988)) and unconditionally by `MemPoolAccept::PreChecks` ([source](https://github.com/dashpay/dash/blob/e6088d4e5f8f4c930ca66ad579237df733b35bc2/src/validation.cpp#L712)), to account for the second usage, we add an assertion.

File: src/governance/governance.cpp (hunk 1 above)
Comment (kwvg) [reply src/governance/governance.cpp:400]: These changes are introduced in https://github.com/dashpay/dash/pull/5966 and have been remedied over there.

**Comment by kwvg** [comment]
Marked as draft to avoid confusion with dependency PR

**Review by PastaPastaPasta — APPROVED** [review]
utACK a5be37c58b729c2d3558178bd1fc8f917e7395c0

**Review by knst — APPROVED** [review]
LGTM a5be37c58b729c2d3558178bd1fc8f917e7395c0

**Review by UdjinM6 — APPROVED** [review]
utACK

**Pull request merged**